secure/storage/github/src/lib.rs

```diff
@@ -0,0 +1,289 @@
+                .as_ref()
+                .ok_or_else(|| Error::InternalError("Not content found".into()))?;
+            Ok(content.trim().into())
+        } else {
+            Err(Error::InternalError(format!(
+                "get mismatch, found {} entries",
+                value.len()
+            )))
+        }
+    }
+
+    /// Create or update a file.
+    pub fn put(&self, path: &str, content: &str) -> Result<(), Error> {
+        let json = if let Some(hash) = self.get_sha(path)? {
+            json!({ "content": content, "message": "libra-secure", "sha": hash })
+        } else {
+            json!({ "content": content, "message": "libra-secure" })
+        };
+
+        let resp = self
+            .upgrade_request(ureq::put(&self.url(path)))
+            .send_json(json);
+        match resp.status() {
+            200 => Ok(()),
+            201 => Ok(()),
+            _ => Err(resp.into()),
+        }
+    }
+
+    /// Simple wrapper around requests to add default parameters to the rquest
+    fn upgrade_request(&self, mut request: ureq::Request) -> ureq::Request {
+        request
+            .set("Authorization", &format!("token {}", self.token))
+            .set(ACCEPT_HEADER, ACCEPT_VALUE)
+            .timeout_connect(TIMEOUT);
+        request
+    }
+
+    /// Get can read files or directories, this makes it easier to use
+    fn get_internal(&self, path: &str) -> Result<Vec<GetResponse>, Error> {
+        let resp = self.upgrade_request(ureq::get(&self.url(path))).call();
+        match resp.status() {
+            200 => {
+                let resp = resp.into_string()?;
+                let get_resp: Result<GetResponse, Error> =
+                    serde_json::from_str(&resp).map_err(|e| e.into());
+
+                if let Ok(get_resp) = get_resp {
+                    return Ok(vec![get_resp]);
+                }
+
+                let get_resp: Result<Vec<GetResponse>, Error> =
+                    serde_json::from_str(&resp).map_err(|e| e.into());
+                if let Ok(get_resp) = get_resp {
+                    return Ok(get_resp);
+                }
+
+                Err(Error::SerializationError(resp))
+            }
+            404 => Err(Error::NotFound(self.url(path))),
+            _ => Err(resp.into()),
+        }
+    }
+
+    /// Returns the sha hash of a file according to GitHub
+    fn get_sha(&self, path: &str) -> Result<Option<String>, Error> {
+        match self.get_internal(path) {
+            Ok(value) => {
+                if let Some(value) = value.iter().find(|entry| entry.path == path) {
+                    Ok(Some(value.sha.clone()))
+                } else {
+                    // This is a directory
+                    Err(Error::InternalError(format!(
+                        "get mismatch, found {} entries",
+                        value.len()
+                    )))
+                }
+            }
+            Err(Error::NotFound(_)) => Ok(None),
```

Comment:
Nit: it seems a little odd to me to be distinguishing the case where the file doesn't exist and return Ok(None). My feeling is that get_sha() should really return Result<String, Error> and in the case the file doesn't exist, it should be a NoFound error... but I can see that you use this in the put() function though.. so it's probably okay as-is.

Reply:
yes... as-is it worked, but now I cleaned it up, thanks